### ISSUE TITLE
make denoising step faster (or optional)

### ISSUE THREAD

**Issue by jcohenadad** [neuropoly/spinalcordtoolbox#1059]
some data are extremely long to process (>30min). 

data: **folder_shared/sct_issue/20161203_issue1059/**
p.s. the seg in not in the same space-- you can use sct_register_multimodal -identity 1

**Comment by SaraDupont** [comment]
On my computer, with these data: 
- Denoising within `sct_segment_graymatter` takes ~4min and GM segmentation in total takes from 5'45" to 7'30". 
- Denoising alone (calling `sct_maths`) takes around 7sec. 

--> I will see if I can fasten the denoising within `sct_segment_graymatter` at the speed of `sct_maths -denoise `.

**Comment by jcohenadad** [comment]
Sara, is the denoising done after or before interpolation?  It should be done before (because interpolation changes noise properties in the image)

**Comment by SaraDupont** [comment]
It is done after to have a smaller image (size 75x75xZ voxels). I'll try to do it before and see if the processing time changes a lot. 

For now, changing the denoising parameters in `sct_segment_graymatter` speeded it up to 8sec (total processing time of 1'30" VS 5'45" with old parameters)

**Comment by peristeri** [comment]
It might be easy to find the bottleneck using cprofile. https://docs.python.org/2/library/profile.html @SaraDupont I can help you with using this tool.

**Comment by SaraDupont** [comment]
Ah nice, I didn't know about cprofile, it could be useful to see if other processing steps can be fasten !

**Comment by SaraDupont** [comment]
Working with the original image (size 512x512x30 vox):
- With denoising **after** interpolation with OLD param (image size during denoising: 75x75x30 vox): 
    - denoising: ~4min
    - total segmentation: 7'30"

- With denoising **after** interpolation with NEW param (image size during denoising: 75x75x30 vox): 
    - denoising: ~8sec
    - total segmentation: ~2min

- With denoising **before** interpolation, with NEW param (on full sized image): 
    - denoising: 6'15
    - total segmentation: 8'20"

- With denoising **before** interpolation, with NEW param, with an approximative pre-crop to fasten denoising: 
    - cropping: 10sec
    - denoising: 10sec
    - total segmentation: 2'15"

**Comment by SaraDupont** [comment]
@jcohenadad Do you want to try the new version on the machine that took >30min to process ? 
It's on branch `sd_issue_1059`

**Comment by SaraDupont** [comment]
Computation on djongo on sct_testing/large:
- on branch sd_issue1059:
```
Mean: {'med_dist': 0.38691221747151755, 'duration_[s]': 222.44895440340042, 'dice_wm': 0.893424123284146, 'hausdorff': 1.5994234057373238, 'dice_gm': 0.6207222204994831}
STD: {'med_dist': 0.13866713409208728, 'duration_[s]': 114.44062279917215, 'dice_wm': 0.0229926743781145, 'hausdorff': 0.30379096956509477, 'dice_gm': 0.09775863615469825}

```
- on master
```
Mean: {'med_dist': 0.38480220968878986, 'duration_[s]': 251.94391154050828, 'dice_wm': 0.8983245179943837, 'hausdorff': 1.6033622974563104, 'dice_gm': 0.6280646411108158}
STD: {'med_dist': 0.13785229645211938, 'duration_[s]': 183.3041171181169, 'dice_wm': 0.023033543868876898, 'hausdorff': 0.3050878646592654, 'dice_gm': 0.10029562247192686}

```

--> I'm surprised by the small difference of computation time between master and branch sd_issue1059

**Comment by jcohenadad** [comment]
i'm trying to compare the syntax-- could you please copy the txt files into /sct_testing

**Comment by jcohenadad** [comment]
add violin plot for test function and then discuss results

**Comment by SaraDupont** [comment]
Here is the comparison between master and branch sd_issue_1059, test run on django on sct_testing/large:

![fig_compare](https://cloud.githubusercontent.com/assets/5857909/22216264/6288b57e-e16c-11e6-8526-11de8d20edfd.png)

**Comment by jcohenadad** [comment]
performance looks similar for lower processing time on branch_1059. Woukd it be possible to investigate the origin of >600s processing time on the very few datasets?
thanks

**Comment by SaraDupont** [comment]
Yes, I'm on it !

**Comment by SaraDupont** [comment]
All the data that have >300sec processing time are in high resolution (0.25x0.25x2.5 mm) and have 26 to 28 slices hence the difference of processing time (GMseg challenge data from Zurich). 

When I run one segmentation alone on django (for ex `zurich_gmsegChallenge16_01`) it takes ~150sec, instead of 725 to 870sec in multi-processing. Thus, the multi-processing biases the results. 

--> The computation time depends on the data size, on its resolution and on the "amount" of processes running in parallel. That amount of processes can change during the run of `isct_test_function`. 
To have a representative computation time I will relaunch the same tests without multi-processing on master and on branch _1059.

**Comment by jcohenadad** [comment]
I see, the "multi-processing" option on isct_test_function sets single-core computation for each subject. But some users only have 4-core (or less), so it would be good to investigate where is the bottleneck in computation time. Aren't all data resampled at 0.3mm in-plane? Is the difference due to the number of slices? Does computation time scale quasi-linearly with number of slices?

**Comment by SaraDupont** [comment]
Yes, all data are resampled at 0.3mm in-plane but the resampling part (forward and backward) takes more time for data with a higher resolution.  It takes also more time for larger data (and high res data are often large). 

Here is a plot of the processing time as a function of the number of slices. The color code indicates the in-plane resolution (white is 0.25 and black is 0.5).
![fig_time_f_n_slices](https://cloud.githubusercontent.com/assets/5857909/22264249/4565efdc-e245-11e6-9d8d-2e1c2509f538.png)
--> For the same resolution, the time does scale quasi-linearly with number of slices. 

Here is a plot of the processing time as a function of the in-plane resolution. The color code indicates the number of slices (white is 3 and black is 29).
![fig_time_f_resolution](https://cloud.githubusercontent.com/assets/5857909/22264393/c4152f78-e245-11e6-9cc3-cb484888396f.png)

With these data, it is hard to determine whether the long processing time is due to the high resolution, the large number of slices or the large size of the data, since the data with the very high processing time combine these factors. I can investigate to disentangle the effect of each of these factors on the processing time.

**Comment by jcohenadad** [comment]
i think we have a good idea now that the high processing time on a few datasets was mainly due to the large number of slices, which we don't need to address now. So we can move forward with the merge, thank you for these thorough investigations

**Comment by SaraDupont** [comment]
Just merged PR #1120
